staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go

```diff
@@ -17,77 +17,203 @@ limitations under the License.
 package log
 
 import (
+	"errors"
 	"fmt"
 	"io"
 	"strings"
+	"sync"
+	"time"
 
+	"github.com/golang/glog"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
 	auditinternal "k8s.io/apiserver/pkg/apis/audit"
 	"k8s.io/apiserver/pkg/audit"
 )
 
 const (
+	// ModeBlocking indicates that the backend should process events directly,
+	// Serialize event, then write it to logs. This mode spend more time.
+	ModeBlocking = "blocking"
+	// ModeBuffered indicates that the audit backend should buffer audit events
+	// internally. If the buffer is full, audit event will be dropped, so no
+	// block event producer.
+	ModeBuffered = "buffered"
+)
+
+// AllowedModes is the modes known by this webhook.
+var AllowedModes = []string{
+	ModeBlocking,
+	ModeBuffered,
+}
+
+const (
 	// FormatLegacy saves event in 1-line text format.
 	FormatLegacy = "legacy"
 	// FormatJson saves event in structured json format.
 	FormatJson = "json"
+
+	// TODO: make this configurable
+	// Number of workers that concurrently write events.
+	workers = 5
```

Comment:
Why have workers? What does this optimize?

Reply:
workers are consumers of events.

Reply:
@ericchiang Presumably that can optimize the json serialization time, if it's slower than writing to a file

However, I'm totally opposed to the idea of workers when we have "buffered backend": it will be paralleled even without those workers, since each batch is processed in a separate goroutine

Reply:
> However, I'm totally opposed to the idea of workers when we have "buffered backend": it will be paralleled even without those workers, since each batch is processed in a separate goroutine

Seconded. I'd rather we just avoid the worker routines.

Reply:
Then you all mean here remove preset workers, as webhook batch backend, run a go routine and receive event from channel and then start a new goroutine to process. 
But if use the same as webhook batch backend, still we need flow control to restrict the routine number. And I don't think repetly start a new goroutine->write event->done is more efficient than preset goroutines. Also add complexity of control goroutine number.

Reply:
> we need flow control to restrict the routine number

It's done already by https://github.com/kubernetes/kubernetes/pull/53417

Reply:
Please remove workers

Reply:
ok. keep consistent